bioio_ome_zarr/writers/metadata.py

```diff
@@ -0,0 +1,329 @@
+from dataclasses import dataclass
+from typing import Any, Dict, List, Optional, Sequence, Tuple
+
+# Public NGFF version constants
+OME_NGFF_VERSION_V04 = "0.4"
+OME_NGFF_VERSION_V05 = "0.5"
+
+
+class Axes:
+    """
+    Holds axis metadata for an N-D image, aligned with NGFF 0.5 axes, and
+    renders cleanly for both NGFF 0.4 (Zarr v2) and NGFF 0.5 (Zarr v3).
+
+    Attributes
+    ----------
+    ndim : int
+        Number of dimensions in the image data.
+    names : List[str]
+        Names of each axis (e.g., ["t", "c", "z", "y", "x"]).
+    types : List[str]
+        NGFF axis types for each axis (e.g., "time", "channel", "space").
+    units : List[Optional[str]]
+        Physical units for each axis, if any (e.g., "micrometer").
+    scales : List[float]
+        Physical scale at level 0 per axis (e.g., pixel sizes).
+    factors : Tuple[int, ...]
+        Per-axis downsample factor (not required by NGFF; useful for clients).
+    """
+
+    DEFAULT_NAMES: List[str] = ["t", "c", "z", "y", "x"]
+    DEFAULT_TYPES: List[str] = ["time", "channel", "space", "space", "space"]
+    DEFAULT_UNITS: List[Optional[str]] = [None, None, None, None, None]
+    DEFAULT_SCALES: List[float] = [1.0, 1.0, 1.0, 1.0, 1.0]
+
+    def __init__(
+        self,
+        *,
+        ndim: int,
+        names: Optional[List[str]] = None,
+        types: Optional[List[str]] = None,
+        units: Optional[List[Optional[str]]] = None,
+        scales: Optional[List[float]] = None,
+        factors: Tuple[int, ...] = (),
+    ) -> None:
+        self.ndim = int(ndim)
+        self.names = list(
+            names[-ndim:] if names is not None else self.DEFAULT_NAMES[-ndim:]
+        )
+        self.types = list(
+            types[-ndim:] if types is not None else self.DEFAULT_TYPES[-ndim:]
+        )
+        self.units = list(
+            units[-ndim:] if units is not None else self.DEFAULT_UNITS[-ndim:]
+        )
+        self.scales = list(
+            scales[-ndim:] if scales is not None else self.DEFAULT_SCALES[-ndim:]
+        )
+        self.factors = tuple(factors[-ndim:] if factors else (1,) * self.ndim)
+
+        if not (
```

Comment:
doesn't the [-ndim:] operation mean these values MUST be the same length as ndim?  I suggest removing the slice operation, and leaving the exception, so we don't silently drop array elements.